File: server/seed-root-user.js
/* ==========================================
   Knowledge Repository - Seed Root User
   Ensures rondayan42@gmail.com exists and is admin/root
   ========================================== */

require('dotenv').config();
const supabase = require('./supabase');

const ROOT_EMAIL = 'rondayan42@gmail.com';
const ROOT_PASSWORD = process.env.ROOT_INITIAL_PASSWORD || 'AdminRoot2025!';

async function seedRootUser() {
    console.log(`Checking for root user: ${ROOT_EMAIL}...`);

    try {
        // 1. Check if user exists (by listing users, limited to 1 filtered by email if possible, 
        // but admin.listUsers doesn't filter by email directly in all versions, 
        // so we'll try to find them manually or use create and catch error)

        let userId = null;
        let userExists = false;

        // Note: listUsers is paginated, but for a specific email search we might need a different approach or 
        // just try to create and handle conflict. However, let's try to find them first to avoid password reset if present.

        // Better approach: Try to get user by email directly if the SDK supports it, or iterate.
        // The standard admin SDK usually allows listUsers.
        const { data: { users }, error: listError } = await supabase.auth.admin.listUsers();

        if (listError) throw listError;

        const existingUser = users.find(u => u.email === ROOT_EMAIL);

        if (existingUser) {
            console.log('Root user found.');
            userId = existingUser.id;
            userExists = true;
        }

        if (!userExists) {
            console.log('Root user not found. Creating...');
            const { data, error: createError } = await supabase.auth.admin.createUser({
                email: ROOT_EMAIL,
                password: ROOT_PASSWORD,
                email_confirm: true,
                user_metadata: {
                    role: 'admin',
                    is_root: true
                }
            });

            if (createError) throw createError;
            console.log(`Root user created successfully. ID: ${data.user.id}`);
            console.log('IMPORTANT: Verify the initial password is changed upon first login.');
        } else {
            console.log('Updating root user metadata to ensure admin privileges...');
            const { data, error: updateError } = await supabase.auth.admin.updateUserById(userId, {
                user_metadata: {
                    role: 'admin',
                    is_root: true
                }
            });

            if (updateError) throw updateError;
            console.log('Root user privileges confirmed.');
        }

    } catch (error) {
        console.error('Error seeding root user:', error.message);
        process.exit(1);
    }
}

seedRootUser();

File: server/seed_root_user.py
"""
Seed Root User for Knowledge Repository
Creates or updates the root admin user
"""

import os
import sys

# Add server directory to path
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from dotenv import load_dotenv
load_dotenv()

from db import query, init_database, seed_default_data
from auth import hash_password

ROOT_EMAIL = 'rondayan42@gmail.com'
ROOT_PASSWORD = os.getenv('ROOT_INITIAL_PASSWORD', 'BsmartRoot2025!')


def seed_root_user():
    print(f'Checking for root user: {ROOT_EMAIL}...')
    
    try:
        # Initialize database first
        init_database()
        seed_default_data()
        
        # Check if root user exists
        result = query('SELECT * FROM users WHERE email = %s', (ROOT_EMAIL,))
        
        if result['rows']:
            print('Root user already exists.')
            # Ensure admin privileges
            query(
                'UPDATE users SET role = %s, approved = %s, is_root = %s WHERE email = %s',
                ('admin', True, True, ROOT_EMAIL)
            )
            print('Root user privileges confirmed.')
        else:
            print('Root user not found. Creating...')
            password_hash = hash_password(ROOT_PASSWORD)
            
            query(
                'INSERT INTO users (email, password_hash, role, approved, is_root) VALUES (%s, %s, %s, %s, %s)',
                (ROOT_EMAIL, password_hash, 'admin', True, True)
            )
            
            print(f'Root user created successfully.')
            print(f'Email: {ROOT_EMAIL}')
            print(f'Password: {ROOT_PASSWORD}')
            print('IMPORTANT: Change the default password after first login.')
        
    except Exception as e:
        print(f'Error seeding root user: {e}')
        sys.exit(1)


if __name__ == '__main__':
    seed_root_user()
